### ISSUE TITLE
some questions about sequence parallel

### ISSUE THREAD

**Issue by yingtongxiong** [NVIDIA/Megatron-LM#357]
Hi, I have some questions about sequence parallel.
<img width="661" alt="捕获" src="https://github.com/NVIDIA/Megatron-LM/assets/45058324/0f0d1176-e03f-43dd-8d30-5e0e38d55123">
I am very confused about the last sequence parallel in transformer when I read your code. As we can see in this [https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/transformer.py#L152](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/transformer.py#L152,) the last part in MLP, which is a rowParallel. If sequence parallel set to be True, there is a reduce and scatter operation in RowParallel [https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/layers.py#L707](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/layers.py#L707). I consider this operation is prepared for the sequence parallel in the dropout after MLP as the above picture shows. However, I don't find any all-gather operation after the dropout and the transformer layer output is sub-sequence as you can see in this [https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/transformer.py#L896](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/transformer.py#L896). Therefore, I am very confused about how can I get the entire transformer output(the output shape is (s, b, h)) and where is the code doing the all-gather operation. Could you help me with this？

**Comment by eyuansu62** [comment]
It may be do the all reduce operations before dropout?

**Comment by yingtongxiong** [comment]
Thank you. I have got it. The all-gather operation will be performed before or after calculating logits if sequence_parallel is True.

**Comment by jon-barker** [comment]
Closing - question appears to have been answered.

**Comment by zte-tcb** [comment]
all-gather in
forward pass and reduce-scatter in backward pass

**Comment by zte-tcb** [comment]
You can find do all-gather in forward pass and reduce-scatter in backward pass in ColumnParallelLinear, and do reduce-scatter in forward pass and all-gather in backward pass in RowParallelLinear.  ColumnParallelLinear is exactly the entrance layer of Attention and MLP. RowParallelLinearis exactly the exit layer of Attention and MLP.

**Comment by tonyz0x0** [comment]
More specifically the input will be gathered in `linear_with_frozen_weight()`:
```
    if sequence_parallel:
        input = gather_from_sequence_parallel_region(input, tensor_parallel_output_grad=True)
    else:
        input = input
```
